### ISSUE TITLE
DoubleTensor test fails on all architectures

### ISSUE THREAD

**Issue by borisfom** [soumith/cudnn.torch#226]
I have checked with caffe guys and they say there is no way those errors may have numeric origin (when they change float to double in their networks, it never increases the error).
So I think, it's reasonable to assume we may have some layout/alignment/copy error in templated types handling, at least for double type.
The error values also differ by order of magnitude between runs, which also suggest reading garbage or corruption:

SpatialFullConvolution_backward_batch
error on bias (backward)
LT failed: 931.37835693359 >= 0.02
        test/test.lua:348: in function 'test'
        test/test.lua:351: in function <test/test.lua:301>

... next run: 

SpatialFullConvolution_backward_batch
error on bias (backward)
LT failed: 94.976547241211 >= 0.02
        test/test.lua:348: in function 'test'
        test/test.lua:351: in function <test/test.lua:301>

**Comment by szagoruyko** [comment]
@borisfom indeed, the outputs of convolution in double contain the same small number, and there are a few places were extraBuffer size is allocated with "4" sizeof hardcoded, however fixing this doesn't help. I think there is another issue somewhere in the bindings.

**Comment by borisfom** [comment]
Allocating with size 4 is indeed a problem (as it's a LongTensor), I have fixed it in my clone as well to always use elementSIze and yes it did not help.
From time to time I am getting segmentation faults (usually when tweaking CUDNN params incorrectly, but still, should not crash). I think it's worth checking memory, too. I have tried  running valgrind on torch, but luajit has issues with it even if rebuilt with dedicated VALGRIND flag. Did you try it before? What would be the workarounds - build with Lua instead ?
@soumith  : any hints ?

**Comment by borisfom** [comment]
Those are my results now. Note all 'functional' tests pass, could there be a clue here?

th ./cudnn/test/test.lua
Running test on device: #1 : Graphics Device
Testing torch.CudaDoubleTensor
Running 38 tests
 1/38 VolumetricMaxPooling_batch ........................................ [PASS]
 2/38 SpatialMaxPooling_single .......................................... [FAIL]
 3/38 SpatialCrossEntropyCriterion ...................................... [PASS]
 4/38 SpatialMaxPooling_batch ........................................... [PASS]
 5/38 SpatialAveragePooling_batch ....................................... [PASS]
 6/38 LogSoftMax_single ................................................. [FAIL]
 7/38 functional_bias2D ................................................. [PASS]
 8/38 SpatialBatchNormalization ......................................... [FAIL]
 9/38 SpatialConvolution_forward_single ................................. [FAIL]
10/38 functional_convolution2d .......................................... [PASS]
11/38 SpatialFullConvolution_backward_batch ............................. [FAIL]
12/38 TemporalConvolution_padding_batch ................................. [FAIL]
13/38 SpatialConvolution_backward_single ................................ [FAIL]
14/38 Tanh_single ....................................................... [FAIL]
15/38 ClippedReLU_single ................................................ [PASS]
16/38 ReLU_single ....................................................... [FAIL]
17/38 Sigmoid_single .................................................... [FAIL]
18/38 VolumetricConvolution_forward_single .............................. [FAIL]
19/38 Tanh_batch ........................................................ [FAIL]
20/38 SoftMax_batch ..................................................... [FAIL]
21/38 SpatialCrossMapLRN_batch .......................................... [FAIL]
22/38 VolumetricMaxPooling_single ....................................... [FAIL]
23/38 VolumetricBatchNormalization ...................................... [FAIL]
24/38 ReLU_batch ........................................................ [FAIL]
25/38 SpatialConvolution_backward_batch ................................. [FAIL]
26/38 TemporalConvolution_batch ......................................... [FAIL]
27/38 SpatialFullConvolution_forward_batch .............................. [FAIL]
28/38 Sigmoid_batch ..................................................... [FAIL]
29/38 SpatialLogSoftMax ................................................. [FAIL]
30/38 TemporalConvolution_reduceBatchSize ............................... [PASS]
31/38 LogSoftMax_batch .................................................. [FAIL]
32/38 BatchNormalization ................................................ [FAIL]
33/38 TemporalConvolution_single ........................................ [FAIL]
34/38 functional_maxpooling2d ........................................... [PASS]
35/38 SpatialConvolution_forward_batch .................................. [FAIL]
36/38 SpatialAveragePooling_single ...................................... [FAIL]
37/38 VolumetricConvolution_backward_single ............................. [FAIL]
38/38 SoftMax_single .................................................... [FAIL]
Completed 253 asserts in 38 tests with 29 failures and 0 errors

**Comment by szagoruyko** [comment]
@borisfom functional tests are not type specific, they test only float.
I checked with my RTC version of SpatialConvolution for double that it works correctly, so I think it's more likely that there is an issue in the bindings than in cutorch. @soumith I think we need your help here, any ideas?

**Comment by ngimel** [comment]
After updating torch, with double I have intermittent errors with SpatialMaxPooling_single and intermittent illegal memory access in SpatialConvolution_forward_single, the rest of the tests pass.

**Comment by borisfom** [comment]
Do you mean updating it as per our discussion in https://github.com/soumith/cudnn.torch/pull/228 ?

**Comment by ngimel** [comment]
I've updated all packages (./update.sh && ./install.sh), and pulled Sergey's buffer-sizeof branch of cudnn.torch. Now after I've fixed extraBuffer size according to my comment in 228 I can not longer get SpatialConvolution to error with illegal memory access, but may be I'm just lucky.

**Comment by borisfom** [comment]
That makes perfect sense! I will fix my refactoring change accordingly and push the change.

**Comment by borisfom** [comment]
@ngimel : somehow doing similar changes in my workspace did not fix double tests for me - can you push your changes to your clone so I can peek ?

**Comment by borisfom** [comment]
@ngimel: mystery solved: I forgot to mention that you need to modify tolerance in test.lua for double tests (make them the same as float) to unmask errors. I have printed some outputs for debug and it looks like zeros are being copied instead of double values.

**Comment by soumith** [comment]
I've fixed all double tensor issues. Now all tests pass for Float and Double Tensor. 

For Half, the tests only dont pass for *Convolution's backward, because CuDNN does not support all filter shapes / sizes for half yet. 
All other tests pass for Half.
